signature/jws/plugin.go

```diff
@@ -0,0 +1,185 @@
+// by delegating the one or both operations to the named plugin,
+// as defined in
+// https://github.com/notaryproject/notaryproject/blob/main/specs/plugin-extensibility.md#signing-interfaces.
+type PluginSigner struct {
+	Runner       plugin.Runner
+	KeyID        string
+	PluginConfig map[string]string
+}
+
+// Sign signs the artifact described by its descriptor, and returns the signature.
+func (s *PluginSigner) Sign(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	out, err := s.Runner.Run(ctx, plugin.CommandGetMetadata, nil)
+	if err != nil {
+		return nil, fmt.Errorf("metadata command failed: %w", err)
+	}
+	metadata, ok := out.(*plugin.Metadata)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect get-plugin-metadata response type '%T'", out)
+	}
+	if err := metadata.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid plugin metadata: %w", err)
+	}
+	if metadata.HasCapability(plugin.CapabilitySignatureGenerator) {
+		return s.generateSignature(ctx, desc, opts)
+	} else if metadata.HasCapability(plugin.CapabilityEnvelopeGenerator) {
+		return s.generateSignatureEnvelope(ctx, desc, opts)
+	}
+	return nil, fmt.Errorf("plugin does not have signing capabilities")
+}
+
+func (s *PluginSigner) describeKey(ctx context.Context, config map[string]string) (*plugin.DescribeKeyResponse, error) {
+	req := &plugin.DescribeKeyRequest{
+		ContractVersion: "1",
+		KeyID:           s.KeyID,
+		PluginConfig:    config,
+	}
+	out, err := s.Runner.Run(ctx, plugin.CommandDescribeKey, req)
+	if err != nil {
+		return nil, fmt.Errorf("describe-key command failed: %w", err)
+	}
+	resp, ok := out.(*plugin.DescribeKeyResponse)
+	if !ok {
+		return nil, fmt.Errorf("plugin runner returned incorrect describe-key response type '%T'", out)
+	}
+	return resp, nil
+}
+
+func (s *PluginSigner) generateSignature(ctx context.Context, desc signature.Descriptor, opts notation.SignOptions) ([]byte, error) {
+	config := s.mergeConfig(opts.PluginConfig)
+	// Get key info.
+	key, err := s.describeKey(ctx, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check keyID is honored.
+	if s.KeyID != key.KeyID {
+		return nil, fmt.Errorf("keyID mismatch")
+	}
+
+	// Get algorithm associated to key.
+	alg := key.KeySpec.SignatureAlgorithm()
+	if alg == "" {
+		return nil, fmt.Errorf("keySpec %q not supported: ", key.KeySpec)
+	}
+
+	// Generate payload to be signed.
+	payload := packPayload(desc, opts)
+	if err := payload.Valid(); err != nil {
+		return nil, err
+	}
+
+	// Generate signing string.
+	token := jwtToken(alg.JWS(), payload)
+	signing, err := token.SigningString()
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal signing payload: %v", err)
+	}
+
+	// Execute plugin sign command.
+	req := &plugin.GenerateSignatureRequest{
+		ContractVersion: "1",
+		KeyID:           s.KeyID,
+		KeySpec:         key.KeySpec,
+		Hash:            alg.Hash(),
+		Payload:         signing,
```

Comment:
Is this base64 encoded a second time? Updated the spec as per your [suggestion](https://github.com/notaryproject/notaryproject/pull/155#discussion_r871991310).

Reply:
Done!